lib/azure/armrest/model/hash_model.rb

```diff
@@ -0,0 +1,140 @@
+require 'active_support/core_ext/string/inflections'
+
+module Azure
+  module Armrest
+    class HashModel
+      class << self
+        private
+
+        # Convert a key to a model class attribute which must be a qualified ruby method name
+        # key is from the original JSON or Hash object
+        #
+        # This is the default implementation, to be overridden by subclasses.
+        #
+        def key_to_attribute(key)
+          key.to_s.underscore
+        end
+
+        # Should setter methods be automatically created?
+        #
+        # Default to true, to be overridden by subclasses.
+        #
+        def generate_writer?
+          true
+        end
+
+        # Should values of this attribute be excluded from converting to model objects
+        #
+        # Default to false, to be overridden by subclasses.
+        #
+        def attr_excluded?(_attr_name)
+          false
+        end
+
+        # Declare attributes that need to be converted to Model
+        def attr_model(*attrs)
+          # Merge the declared attributes to the modelized list
+          @modelized_list = modelized_list | Set.new(attrs.map(&:to_s))
+        end
+
+        # Return the list of modelized attributes
+        def modelized_list
+          @modelized_list ||= Set.new
+        end
+
+        def convert_value(key, value, parent)
+          method = key_to_attribute(key)
+          new_val =
+            if attr_excluded?(method)
+              value
+            else
+              # Must deal with nested models
+              case value
+              when Array
+                value.collect do |elem|
+                  elem.kind_of?(Hash) ? parent.send(:hash_to_model, method.camelize.singularize, elem) : elem
+                end
+              when Hash
+                parent.send(:hash_to_model, method.camelize, value)
+              else
+                value
+              end
+            end
+
+          add_accessor_methods(method, key)
+          new_val
+        end
+
+        def add_accessor_methods(method, key)
+          return if modelized_list.include?(method.to_s)
+
+          method = "_#{method}" if instance_methods.include?(method.to_sym)
+          class_eval { define_method(method) { @data[key] } }
+          attr_model(method)
+
+          return unless generate_writer?
+          class_eval { define_method("#{method}=") { |val| @data[key] = val } }
+        end
+      end ## self
+
+      def initialize(json_or_hash)
+        raw_hash = json_or_hash.kind_of?(Hash) ? json_or_hash : JSON.parse(json_or_hash)
+        @data = Hash[raw_hash.collect { |key, value| [key, self.class.send(:convert_value, key, value, self)] }]
```

Comment:
Is there a reason we are going from a `Hash`, to an `Array`, and back again here?  Couldn't this just be simplified to:

```ruby
def initialize(json_or_hash)
  @data = json_or_hash.kind_of?(Hash) ? json_or_hash : JSON.parse(json_or_hash)
  @data.each {|key, value| @data[key] = self.class.send(:convert_value, key, value, self) }
end
```